src/main/flight/interpolated_setpoint.c

```diff
@@ -0,0 +1,143 @@
+ * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this software.
+ *
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "platform.h"
+#include "interpolated_setpoint.h"
+#include "build/debug.h"
+#include "fc/rc.h"
+#include "common/maths.h"
+#include "math.h"
+
+#ifdef USE_INTERPOLATED_SP
+static float projectedSetpoint[XYZ_AXIS_COUNT];
+static float prevSetpointSpeed[XYZ_AXIS_COUNT];
+static float prevRawSetpoint[XYZ_AXIS_COUNT];
+static float prevRawDeflection[XYZ_AXIS_COUNT];
+static uint16_t interpolationSteps[XYZ_AXIS_COUNT];
+static float setpointChangePerIteration[XYZ_AXIS_COUNT];
+static float deflectionChangePerIteration[XYZ_AXIS_COUNT];
+static float setpointReservoir[XYZ_AXIS_COUNT];
+static float deflectionReservoir[XYZ_AXIS_COUNT];
+
+// Configuration
+static float ffLookaheadLimit;
+static float ffSpread;
+static float ffMaxRateLimit[XYZ_AXIS_COUNT];
+static float ffMaxRate[XYZ_AXIS_COUNT];
+
+void interpolatedSpInit(const pidProfile_t *pidProfile) {
+    const float ffMaxRateScale = pidProfile->ff_max_rate_limit * 0.01f;
+    ffLookaheadLimit = pidProfile->ff_lookahead_limit * 0.0001f;
+    ffSpread = pidProfile->ff_spread;
+    for (int i = 0; i < XYZ_AXIS_COUNT; i++) {
+        ffMaxRate[i] = applyCurve(i, 1.0f);
+        ffMaxRateLimit[i] = ffMaxRate[i] * ffMaxRateScale;
+    }
+}
+
+FAST_CODE_NOINLINE float interpolatedSpApply(int axis, float pidFrequency, bool newRcFrame) {
+    const float rawSetpoint = getRawSetpoint(axis);
+    const float rawDeflection = getRawDeflection(axis);
+
+    float pidSetpointDelta = 0.0f;
+    static int iterationsSinceLastUpdate[XYZ_AXIS_COUNT];
+    if (newRcFrame) {
+
+        setpointReservoir[axis] -= iterationsSinceLastUpdate[axis] * setpointChangePerIteration[axis];
+        deflectionReservoir[axis] -= iterationsSinceLastUpdate[axis] * deflectionChangePerIteration[axis];
+        iterationsSinceLastUpdate[axis] = 0;
+
+        // get the number of interpolation steps either dynamically based on RX refresh rate
+        // or manually based on ffSpread configuration property
+        if (ffSpread) {
+            interpolationSteps[axis] = (uint16_t) ((ffSpread + 1.0f) * 0.001f * pidFrequency);
+        } else {
+            interpolationSteps[axis] = (uint16_t) ((currentRxRefreshRate + 1000) * pidFrequency * 1e-6f + 0.5f);
+        }
+
+        // interpolate stick deflection
+        deflectionReservoir[axis] += rawDeflection - prevRawDeflection[axis];
+        deflectionChangePerIteration[axis] = deflectionReservoir[axis] / interpolationSteps[axis];
+        const float projectedStickPos =
+            rawDeflection + deflectionChangePerIteration[axis] * pidFrequency * ffLookaheadLimit;
+        projectedSetpoint[axis] = applyCurve(axis, projectedStickPos);
```

Comment:
only `projectedSetpoint[axis] - rawSetpoint [axis]` is actually used. It may be better to use projectedSetpointDelta ..

Reply:
@ledvinap I'm not yet sure how this will evolve so I'll keep it this way if you don't mind. This is more easily understood and modified.